File: python/20_有效的括号_easy.py
class Solution:
    def isValid(self, s: str) -> bool:
        if len(s) % 2 == 1:
            return False
        stack = []
        strMap = {
            ')': '(',
            ']': '[',
            '}': '{'
        }
        for _s in s:
            if _s in strMap:
                if not stack or stack[-1] != strMap[_s]:
                    return False
                stack.pop()
            else:
                stack.append(_s)
        return True if not stack else False

File: cpp/20_有效的括号_easy.cpp
class Solution {
public:
    bool isValid(string s) {
        if (s.length() % 2 == 1)
            return false;
        vector<char> stack;
        unordered_map<char, char> strMap = {
            {')', '('},
            {']', '['},
            {'}', '{'}
        };
        for (char ch : s){
            if (strMap.count(ch) != 0){
                if (!stack.size() || stack[stack.size()-1] != strMap[ch])
                    return false;
                stack.pop_back();
            }
            else
                stack.push_back(ch);
        }
        return stack.size() == 0 ? true : false;
    }
};